assemble/Particle_Diagnostics.F90

```diff
@@ -0,0 +1,1347 @@
+                   temp_part%old_fields(:) = weighted_old_fields(i)%col(:)
+                end if
+
+                node_part_count(node_num) = node_part_count(node_num) + 1
+
+                !add particle to relevant particle list
+                call insert(temp_part, particle_lists(group_arrays(i)))
+
+                !add particle to temp particle list
+                call temp_list_insert(temp_part,node_particles(i,node_num))
+
+                add_particles(i) = add_particles(i) + 1
+                id = id + 1
+                particle_lists(group_arrays(i))%proc_part_count = particle_lists(group_arrays(i))%proc_part_count + 1
+             end do
+          end do
+       end do
+    end do
+
+    allocate(remove_particles(size(group_arrays)))
+    remove_particles = 0
+
+    !Call delete_particles if number of particles now exceeds the maximum threshold
+    do while(node_part_count(node_num)>max_thresh)
+       call delete_particles(node_part_count(node_num), node_particles(:,node_num), group_arrays, remove_particles)
+       add_particles(:) = add_particles(:) - remove_particles(:)
+    end do
+
+    deallocate(remove_particles)
+    deallocate(node_coord)
+    deallocate(node_numbers)
+    deallocate(node_loc)
+    deallocate(ele_val)
+    if (allocated(weighted_attributes)) then
+       deallocate(weighted_attributes)
+       deallocate(weighted_old_attributes)
+       deallocate(weighted_old_fields)
+       deallocate(length_group)
+    end if
+
+  end subroutine spawn_zero_particles
+
+  subroutine set_spawned_lcoords(max_lcoord, node_coord, node_num, node_numbers)
+    !Subroutine to randomly set spawned particle local coordinates based off
+    !the maximum local coordinate given
+
+    real, intent(inout) :: max_lcoord
+    real, dimension(:), intent(inout) :: node_coord
+    integer, intent(in) :: node_num
+    integer, dimension(:), intent(in) :: node_numbers
+
+    integer, dimension(4,4) :: permutation = reshape([1,2,3,4, 2,1,3,4, 2,3,1,4, 2,3,4,1], [4,4])
+    real, dimension(4) :: work
+    real :: tmp_res, rand_val
+    integer :: i, j
+
+    max_lcoord = max(0.51, min(max_lcoord, 0.999))
+    ! set up the node coordinates to be permuted
+    ! looks like: [x, (1 - x) * rand(), (1 - x - y) * rand(), 1 - x - y - z]
+    ! depending on the number of coordinates
+    work(1) = max_lcoord
+    tmp_res = 1 - max_lcoord
+    do j = 2, size(node_coord) - 1
+       call random_number(rand_val)
+       work(j) = tmp_res * rand_val
+       tmp_res = tmp_res - work(j)
+    end do
+    work(size(node_coord)) = tmp_res
+    do i = 1, size(node_coord)
+       ! determine the node index corresponding to the
+       ! target node number
+       if (node_num /= node_numbers(i)) cycle
+
+       ! set the coordinates according to the permutation for this index
+       do j = 1, size(node_coord)
+          node_coord(j) = work(permutation(j,i))
+       end do
+
+    end do
+
```

Comment:
This is already complicated by the permutation, but maybe you can simplify a bit, showing that this is really just two loops consecutively:
```
  do i=1, size(node_coord)
    if (node_num == node_numbers(i)) exit
  end do
  assert(i<=size(node_coord)
  do j=1, size(node_coord)
      node_coord(j) = work(permutation(j,i))
  end do
```